File: objectivec/day18_part2_2018.m

#import <Foundation/Foundation.h>

static const char OPEN='.';
static const char TREES='|';
static const char LUMBERYARD='#';

static char **readInput(NSString *path, NSInteger *rows, NSInteger *cols){
    NSString *content=[NSString stringWithContentsOfFile:path encoding:NSUTF8StringEncoding error:nil];
    NSArray<NSString *> *lines=[content componentsSeparatedByString:@"\n"];
    if(lines.lastObject.length==0) lines=[lines subarrayWithRange:NSMakeRange(0, lines.count-1)];
    *rows=lines.count;
    *cols=lines[0].length;
    char **grid=malloc(sizeof(char*)*(*rows));
    for(NSInteger i=0;i<*rows;i++){
        grid[i]=malloc(sizeof(char)*(*cols));
        const char *cstr=[lines[i] UTF8String];
        memcpy(grid[i], cstr, *cols);
    }
    return grid;
}

static NSInteger countAdjacent(char **g, NSInteger r, NSInteger c, NSInteger rows, NSInteger cols, char type){
    NSInteger cnt=0;
    for(NSInteger dr=-1;dr<=1;dr++)for(NSInteger dc=-1;dc<=1;dc++){
        if(dr==0&&dc==0)continue;
        NSInteger nr=r+dr,nc=c+dc;
        if(nr>=0&&nr<rows&&nc>=0&&nc<cols&&g[nr][nc]==type)cnt++;
    }
    return cnt;
}

static char nextState(char **g, NSInteger r, NSInteger c, NSInteger rows, NSInteger cols){
    char cur=g[r][c];
    if(cur==OPEN) return countAdjacent(g,r,c,rows,cols,TREES)>=3?TREES:OPEN;
    if(cur==TREES) return countAdjacent(g,r,c,rows,cols,LUMBERYARD)>=3?LUMBERYARD:TREES;
    return (countAdjacent(g,r,c,rows,cols,LUMBERYARD)>=1 && countAdjacent(g,r,c,rows,cols,TREES)>=1)?LUMBERYARD:OPEN;
}

static char **transform(char **g, NSInteger rows, NSInteger cols){
    char **ng=malloc(sizeof(char*)*rows);
    for(NSInteger i=0;i<rows;i++){
        ng[i]=malloc(sizeof(char)*cols);
        for(NSInteger j=0;j<cols;j++) ng[i][j]=nextState(g,i,j,rows,cols);
    }
    return ng;
}

static NSString *gridKey(char **g, NSInteger rows, NSInteger cols){
    NSMutableData *data=[NSMutableData dataWithLength:rows*cols];
    for(NSInteger i=0;i<rows;i++) memcpy(data.mutableBytes+i*cols, g[i], cols);
    return [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
}

static void countResources(char **g, NSInteger rows, NSInteger cols, NSInteger *wooded, NSInteger *lumber){
    *wooded=*lumber=0;
    for(NSInteger i=0;i<rows;i++)for(NSInteger j=0;j<cols;j++){
        if(g[i][j]==TREES) (*wooded)++;
        else if(g[i][j]==LUMBERYARD) (*lumber)++;
    }
}

int main(int argc, const char * argv[]) {
    @autoreleasepool{
        NSInteger rows,cols;
        char **grid=readInput(@"input.txt",&rows,&cols);
        NSMutableDictionary<NSString*,NSNumber*> *seen=[NSMutableDictionary dictionary];
        NSInteger minute=0,cycleStart=0,cycleLen=0;
        while(YES){
            NSString *key=gridKey(grid,rows,cols);
            NSNumber *prev=[seen objectForKey:key];
            if(prev){
                cycleStart=prev.integerValue;
                cycleLen=minute-cycleStart;
                break;
            }
            seen[key]=@(minute);
            char **next=transform(grid,rows,cols);
            for(NSInteger i=0;i<rows;i++) free(grid[i]);
            free(grid);
            grid=next;
            minute++;
        }
        NSInteger remaining=(1000000000LL-cycleStart)%cycleLen;
        for(NSInteger i=0;i<remaining;i++){
            char **next=transform(grid,rows,cols);
            for(NSInteger r=0;r<rows;r++) free(grid[r]);
            free(grid);
            grid=next;
        }
        NSInteger wooded,lumber;
        countResources(grid,rows,cols,&wooded,&lumber);
        printf("%lld\n",(long long)(wooded*lumber));
        for(NSInteger i=0;i<rows;i++) free(grid[i]);
        free(grid);
    }
    return 0;
}

File: fsharp/day18_part2_2018.fs
open System
open System.IO
open System.Collections.Generic
open System.Text

let openChar = '.'
let treesChar = '|'
let lumberChar = '#'

let readInput (filename:string) =
    File.ReadAllLines(filename) |> Array.map (fun l -> l.ToCharArray())

let countAdjacent (grid:char[][]) i j target =
    let mutable cnt = 0
    for dx = -1 to 1 do
        for dy = -1 to 1 do
            if not (dx = 0 && dy = 0) then
                let ni = i + dx
                let nj = j + dy
                if ni >= 0 && ni < grid.Length && nj >= 0 && nj < grid.[0].Length && grid.[ni].[nj] = target then
                    cnt <- cnt + 1
    cnt

let nextState (grid:char[][]) i j =
    match grid.[i].[j] with
    | '.' -> if countAdjacent grid i j treesChar >= 3 then treesChar else openChar
    | '|' -> if countAdjacent grid i j lumberChar >= 3 then lumberChar else treesChar
    | '#' ->
        if countAdjacent grid i j lumberChar >= 1 && countAdjacent grid i j treesChar >= 1 then lumberChar
        else openChar
    | c -> c

let transform (grid:char[][]) =
    let rows = grid.Length
    let cols = grid.[0].Length
    Array.init rows (fun i -> Array.init cols (fun j -> nextState grid i j))

let countResources (grid:char[][]) =
    let mutable wooded = 0
    let mutable lumber = 0
    for row in grid do
        for c in row do
            if c = treesChar then wooded <- wooded + 1
            elif c = lumberChar then lumber <- lumber + 1
    (wooded, lumber)

let gridToString (grid:char[][]) =
    let sb = StringBuilder()
    for row in grid do
        sb.AppendLine(String(row)) |> ignore
    sb.ToString()

[<EntryPoint>]
let main argv =
    let mutable grid = readInput "input.txt"
    let seen = Dictionary<string,int>()
    let mutable cycleStart = 0
    let mutable cycleLen = 0
    let mutable minute = 0
    let mutable found = false
    while not found do
        let state = gridToString grid
        if seen.ContainsKey(state) then
            cycleStart <- seen.[state]
            cycleLen <- minute - cycleStart
            found <- true
        else
            seen.[state] <- minute
            grid <- transform grid
            minute <- minute + 1
    let remaining = (1000000000 - cycleStart) % cycleLen
    for _ = 0 to remaining - 1 do
        grid <- transform grid
    let (wooded, lumber) = countResources grid
    printfn "%d" (wooded * lumber)
    0